ksonnet-gen/printer/printer.go

```diff
@@ -0,0 +1,345 @@
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+		buf.WriteString(" then ")
+		b, err = p.output(t.BranchTrue)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+		buf.WriteString(" else ")
+		b, err = p.output(t.BranchFalse)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+	case *ast.Index:
+		buf.WriteString(string(*t.Id))
+		buf.WriteString(".")
+		b, err := p.output(t.Target)
+		if err != nil {
+			return nil, err
+		}
+		buf.Write(b)
+	case *ast.Object:
+		buf.WriteString("{")
+
+		p.level++
+		for i, field := range t.Fields {
+			out, err := p.output(field)
+			if err != nil {
+				return nil, err
+			}
+
+			if !t.Inline {
+				buf.Write([]byte{newline})
+				out = p.indent(out)
+			}
+
+			buf.Write(out)
+			if i < len(t.Fields)-1 {
+				buf.Write([]byte{comma})
+				if t.Inline {
+					buf.Write([]byte{blank})
+				}
+			} else if !t.Inline {
+				buf.Write([]byte{comma})
+			}
+		}
+		p.level--
+
+		if !t.Inline {
+			buf.Write([]byte{newline})
+		}
+
+		buf.WriteString("}")
+	case ast.ObjectField:
+		if err := p.handleObjectField(t, &buf); err != nil {
+			return nil, errors.Wrap(err, "handle object field")
+		}
+	case *ast.LiteralString:
+		var quoteType byte
+
+		switch t.Kind {
+		default:
+			return nil, errors.Errorf("unknown string literal kind %#v", t.Kind)
+		case ast.StringDouble:
+			quoteType = doubleQuote
+		case ast.StringSingle:
+			quoteType = singleQuote
+		}
+
+		buf.Write([]byte{quoteType})
+		buf.WriteString(t.Value)
+		buf.Write([]byte{quoteType})
+	case *ast.LiteralNumber:
+		buf.WriteString(t.OriginalString)
+	case *ast.Self:
+		buf.WriteString("self")
+	case *ast.Var:
+		buf.WriteString(string(t.Id))
+	}
+
+	return buf.Bytes(), nil
+}
+
+func bopToString(bop ast.BinaryOp) (string, error) {
+	for k, v := range ast.BopMap {
+		if v == bop {
+			return k, nil
+		}
+	}
+
+	return "", errors.New("unknown binary operation")
+}
+
+func (p *printer) handleApply(t *ast.Apply, buf *bytes.Buffer) error {
+	s, err := extractApply(t.Target)
+	if err != nil {
+		return err
+	}
+	buf.WriteString(s)
+	buf.WriteString("(")
+	b, err := p.output(t.Arguments)
+	if err != nil {
+		return err
+	}
+	buf.Write(b)
+	buf.WriteString(")")
+
+	return nil
+}
+
+func (p *printer) handleObjectField(t ast.ObjectField, buf *bytes.Buffer) error {
+	if t.Comment != nil {
+		isNew := true
+
+		buf.WriteString("//")
+		for _, r := range t.Comment.Text {
+			if r == rune(newline) {
+				buf.Write([]byte{newline})
+				buf.WriteString("//")
+				isNew = true
+				continue
+			}
+			if isNew {
+				buf.Write([]byte{blank})
+				isNew = false
+			}
+
+			buf.Write([]byte{byte(r)})
+		}
+		buf.Write([]byte{newline})
+	}
+
+	var fieldType string
+
+	switch hide := t.Hide; hide {
+	default:
+		return errors.Errorf("unknown Hide type %#v", hide)
+	case ast.ObjectFieldHidden:
+		fieldType = "::"
+	case ast.ObjectFieldVisible:
+	case ast.ObjectFieldInherit:
+		fieldType = ":"
+	}
+
+	switch kind := t.Kind; kind {
+	default:
```

Comment:
What do you think about explicitly erroring out if it's a type we know we don't support, like `ObjectAssert`?

Same might be done for `Node` types, `StringLiteral` types, and so on.